### PR TITLE
[GEN][ZH] Fix crash in AIUpdateInterface::privateMoveAwayFromUnit() when a delayed AICommandParms relates to a then deleted object

### PR DESCRIPTION
**Pull request by Mauller**
- Resolves: #924 

The Issue contains a full list of my workings and findings on for this particular bug.

The problem stems from a race condition in which a pending AI command on an object can end up containing the ID of an object that has already been destroyed.

Many of the private action commands in ```aiupdate.cpp``` already have checks for null pointers in them. Or the object is passed internally into another function that checks for a null pointer.
But the ```privateMoveAwayFrom()``` function did not have a check, this was the cause of the crash in the attached replay in the issue. 

~~I also added a check in the ```privateguardObject()``` function as this also lacked any protection from a null pointer dereference on the passed object.~~

The AI commands are generic and not all fields are populated, so the most sensible place for this fix is within the affected functions themselves.

### CODE REVIEW

File: Generals/Code/GameEngine/Source/GameLogic/Object/Update/AIUpdate.cpp [hunk 1]
```diff
@@ -3074,6 +3074,11 @@ void AIUpdateInterface::privateMoveAwayFromUnit( Object *unit, CommandSourceType
 	{
 		return;
 	}
+
+	// TheSuperHacker @Bugfix Mauller 26/05/2025 Prevent nullptr use when a delayed command contains the id of a destroyed object
```
Comment (xezon) [inline Generals/Code/GameEngine/Source/GameLogic/Object/Update/AIUpdate.cpp]: `@bugfix`

**Comment by Mauller** [comment]
So from some recent replay testing, this appears to be the cause of Four game crashes so far.

**Comment by Mauller** [comment]
Just rebased with main for build testing reaplays

File: Generals/Code/GameEngine/Source/GameLogic/Object/Update/AIUpdate.cpp (hunk 1 above)
Comment (xezon) [reply Generals/Code/GameEngine/Source/GameLogic/Object/Update/AIUpdate.cpp]: "Fix dereferencing null object when a delayed ai command refers to a deleted object. This originally happened when a China Hacker was unpacking and pushed away by a soon-to-be-destroyed object."

File: GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Update/AIUpdate.cpp [hunk 2]
```diff
@@ -4091,6 +4096,10 @@ void AIUpdateInterface::privateGuardTunnelNetwork( GuardMode guardMode, CommandS
  */
 void AIUpdateInterface::privateGuardObject( Object *objectToGuard, GuardMode guardMode, CommandSourceType cmdSource )
 {
+	// TheSuperHacker @Bugfix Mauller 26/05/2025 Prevent nullptr use when a delayed command contains the id of a destroyed object
+	if (!objectToGuard)
```
Comment (xezon) [inline GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Update/AIUpdate.cpp]: ~~I suggest put this in a separate change, because we have no evidence that this causes a crash. Then~~ the comment can be removed above.

Comment (xezon) [reply GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Update/AIUpdate.cpp]: Instead, I suggest to put a comment in `AIUpdateInterface::aiDoCommand` to explain that commands can be delayed and `m_obj`, `m_otherObj`, `m_team` need test for null.

Comment (Mauller) [reply GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Update/AIUpdate.cpp]: Will do, shall update the comments after work and split out ```privateGuardObject()``` and ```privateCommandButtonObject()``` into their own PR.

Comment (xezon) [reply GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Update/AIUpdate.cpp]: Yes either move them to separate change or just remove the comments. Only the actual fix can be be marked as bugfix. The others are just side quests for robustness.

Comment (Mauller) [reply GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Update/AIUpdate.cpp]: fixed

File: Generals/Code/GameEngine/Source/GameLogic/Object/Update/AIUpdate.cpp (hunk 1 above)
Comment (Mauller) [reply Generals/Code/GameEngine/Source/GameLogic/Object/Update/AIUpdate.cpp]: fixed

**Review by xezon — APPROVED** [review]
Safe.
